shuup/discounts/utils.py

```diff
@@ -57,17 +71,145 @@ def get_potential_discounts_for_product(context, product):
+    from django.utils.timezone import now
+
+    now_datetime = now()
+    next_date = now_datetime + datetime.timedelta(days=(abs(weekday - now_datetime.weekday()) % 7))
+    ranges = [
+        next_date.replace(hour=from_hour.hour, minute=from_hour.minute),
+        next_date.replace(hour=to_hour.hour, minute=to_hour.minute)
+    ]
+
+    # the next date is the same as today, let's return also the next week ranges
+    if next_date.date() == now().date():
+        next_week_date = next_date + datetime.timedelta(days=7)
+        ranges.extend([
+            next_week_date.replace(hour=from_hour.hour, minute=from_hour.minute),
+            next_week_date.replace(hour=to_hour.hour, minute=to_hour.minute)
+        ])
+
+    return ranges
+
+
+def bump_price_expiration(shops):
+    """
+    Bump price expiration cache for shops
+
+    :param itetable[int|Shop] shops: list of shops to bump caches
+    """
+    from shuup.core.models import Shop
+    shop_ids = [shop.pk if isinstance(shop, Shop) else int(shop) for shop in shops]
+
+    for shop_id in shop_ids:
+        context_cache.bump_cache_for_item(_get_price_expiration_cache_key(shop_id))
+
+
+def get_price_expiration(context, product):
+    """
+    Returns the price expiration for the product through a UNIX timestamp
+
+    This routine loads all dates that can possibly affect the price of the product in the future.
+
+    After fetching all the event dates, the expiration time will
+        be the minimum datetime that is greater than now:
+
+        expire_on = min(
+            event_date for event_dates in [
+                next_discount_start,
+                next_discount_ends,
+                next_happy_hour_start,
+                next_happy_hour_end,
+                next_availability_exception_start,
+                next_availability_exception_end
+            ]
+            if event_date > now
+        )
+
+    :rtype numbers.Number|None
+    :returns the price expiration time timestamp
+    """
+    cache_params = dict(
+        identifier="price_expiration",
+        item=_get_price_expiration_cache_key(context.shop.pk),
+        context={}
+    )
+
+    if settings.SHUUP_DISCOUNTS_PER_PRODUCT_EXPIRATION_DATES:
+        cache_params["customer"] = getattr(context, "customer", None)
+        cache_params["product"] = product
+
+    key, value = context_cache.get_cached_value(**cache_params)
+    if value is not None:
+        return value
+
+    context_cache_key = "price_expiration_%(shop_id)s" % dict(shop_id=context.shop.pk)
+    if hasattr(context, "context_cache_key"):
+        return getattr(context, context_cache_key)
+
+    from shuup.discounts.models import AvailabilityException, Discount, TimeRange
+
+    if settings.SHUUP_DISCOUNTS_PER_PRODUCT_EXPIRATION_DATES:
+        potential_discounts = get_potential_discounts_for_product(context, product, available_only=False)
+    else:
+        potential_discounts = Discount.objects.filter(shops=context.shop, active=True)
```

Comment:
Should this also check wether the `end_datetime` is None or in future? We don't want to include discounts which dates is in past.